experimental/zerotrie/src/builder/konst/builder.rs

```diff
@@ -0,0 +1,296 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use super::super::branch_meta::BranchMeta;
+use super::super::bytestr::ByteStr;
+use super::store::const_for_each;
+use super::store::ConstArrayBuilder;
+use super::store::ConstLengthsStack;
+use super::store::ConstSlice;
+use crate::error::Error;
+use crate::varint;
+
+/// A low-level builder for ZeroTrieSimpleAscii. Works in const contexts.
+pub(crate) struct ZeroTrieBuilderConst<const N: usize> {
+    data: ConstArrayBuilder<N, u8>,
+}
+
+impl<const N: usize> ZeroTrieBuilderConst<N> {
+    #[cfg(feature = "litemap")]
+    pub fn as_bytes(&self) -> &[u8] {
+        self.data.as_const_slice().as_slice()
+    }
+
+    pub const fn take_or_panic(self) -> [u8; N] {
+        self.data.const_take_or_panic()
+    }
+
+    pub const fn new() -> Self {
+        Self {
+            data: ConstArrayBuilder::new_empty([0; N], N),
+        }
+    }
+
+    #[must_use]
+    const fn prepend_ascii(self, ascii: u8) -> (Self, usize) {
+        if ascii >= 128 {
+            panic!("Non-ASCII not supported in ZeroTrieSimpleAscii");
+        }
+        let data = self.data.const_push_front_or_panic(ascii);
+        (Self { data }, 1)
+    }
+
+    #[must_use]
+    const fn prepend_value(self, value: usize) -> (Self, usize) {
+        let mut data = self.data;
+        let varint_array = varint::write_varint_meta3(value);
+        data = data.const_extend_front_or_panic(varint_array.as_const_slice());
+        data = data.const_bitor_assign(0, 0b10000000);
+        (Self { data }, varint_array.len())
+    }
+
+    #[must_use]
+    const fn prepend_branch(self, value: usize) -> (Self, usize) {
+        let mut data = self.data;
+        let varint_array = varint::write_varint_meta2(value);
+        data = data.const_extend_front_or_panic(varint_array.as_const_slice());
+        data = data.const_bitor_assign(0, 0b11000000);
+        (Self { data }, varint_array.len())
+    }
+
+    #[must_use]
+    const fn prepend_slice(self, s: ConstSlice<u8>) -> (Self, usize) {
+        let mut data = self.data;
+        let mut i = s.len();
+        while i > 0 {
+            data = data.const_push_front_or_panic(*s.get_or_panic(i - 1));
+            i -= 1;
+        }
+        (Self { data }, s.len())
+    }
+
+    #[must_use]
+    const fn prepend_n_zeros(self, n: usize) -> Self {
+        let mut data = self.data;
+        let mut i = 0;
+        while i < n {
+            data = data.const_push_front_or_panic(0);
+            i += 1;
+        }
+        Self { data }
+    }
+
+    const fn bitor_assign_at(self, index: usize, byte: u8) -> Self {
+        let mut data = self.data;
+        data = data.const_bitor_assign(index, byte);
+        Self { data }
+    }
+
+    /// Panics if the items are not sorted
+    pub const fn from_tuple_slice<'a, const K: usize>(
```

Comment:
issue: please document `K` here and elsewhere

Reply:
Yes, this is important; done